admin/vlaudit-stats/vlaudit-stats

```diff
@@ -0,0 +1,668 @@
+
+        self.ips = HitTable(args, 'host')
+        self.pos = HitTable(args, 'vol')
+        self.neg = HitTable(args, 'vol')
+        self.audit_gaps = collections.deque(maxlen=MAX_GAPS)
+
+    def now(self):
+        if self.args.force_time is None:
+            return int(time.time())
+        return self.args.force_time
+
+    def stats(self):
+        return dict(
+            audit=dict(
+                n_msgs=self.n_msgs,
+                n_unreadable=self.n_bad,
+                last_unreadable=self.last_bad,
+
+                n_miss=self.audit_n_miss,
+                gaps=[gap.stats() for gap in self.audit_gaps],
+
+                n_restarts=self.n_restarts,
+                last_restart_ts=self.last_restart_ts,
+
+                n_weird=self.audit_n_weird,
+                last_weird_fromid=self.audit_weird_from,
+                last_weird_toid=self.audit_weird_to,
+                last_weird_fromts=self.audit_weird_fromts,
+                last_weird_tots=self.audit_weird_tots,
+            ),
+            getentry=dict(
+                n_reqs=self.ips.n_hits,
+                host=self.ips.stats(),
+                pos=self.pos.stats(),
+                neg=self.neg.stats(),
+            ),
+        )
+
+    def message(self, msgid, tstamp):
+        self.n_msgs += 1
+
+        if msgid == 0:
+            self.last_msgid = -1
+            self.last_tstamp = None
+            self.restarted(tstamp)
+
+        if self.last_msgid >= 0:
+            # Our msgid should be exactly 1 higher than the last message that
+            # we got. If it's more than 1 higher, we missed some messages; if
+            # it's anything else, that's weird/wrong. (Possibly the vlserver
+            # restarted and we missed some messages.)
+            diff = msgid - self.last_msgid
+            if diff <= 0:
+                self.msgid_weird(self.last_msgid, msgid, self.last_tstamp, tstamp)
+            elif diff > 1:
+                self.msgid_gap(diff - 1, self.last_tstamp, tstamp)
+
+        self.last_msgid = msgid
+        self.last_tstamp = tstamp
+
+    def bad_line(self, line):
+        self.n_bad += 1
+        self.last_bad = self.now()
+
+    def getent(self, req):
+        self.ips.hit(req.host, req)
+        if req.code == 0:
+            self.pos.hit(req.reqvol, req)
+        else:
+            self.neg.hit(req.reqvol, req)
+
+    def restarted(self, tstamp):
+        self.n_restarts += 1
+        self.last_restart = self.now()
+        self.last_restart_ts = tstamp
+
+    def msgid_weird(self, from_id, to_id, from_ts, to_ts):
+        self.audit_n_weird += 1
+        self.audit_weird_from = from_id
+        self.audit_weird_to = to_id
+        self.audit_weird_last = self.now()
+        self.audit_weird_fromts = from_ts
+        self.audit_weird_tots = to_ts
+        warn("msgid went backwards: %d -> %d (audit tstamp %s -> %s)" % (
+             from_id, to_id, from_ts, to_ts))
+
+    def msgid_gap(self, n_miss, from_ts, to_ts):
```

Comment:
I wasn't sure if message "gaps" should even be recorded in the stats, or we should just log them. the should be rare, and it might be easier to just detect them by looking in logs